sql: allow multiple UDFs with in-exact type parameters of the same family

il9ue opened the pull request:
This patch makes it possible for udf to check exact type matching for udf parameters #94146 
Added TypeList.MatchExact to apply udf parameter check only.

Unit test for such family type parameter matching are provided

Release note: None

KaiSun31415 committed:
storage: Refactor resolve write intent options

In this PR, we add MVCCResolveWriteIntentOptions to
MVCCResolveWriteIntent and MVCCResolveWriteIntentRangeOptions to
MVCCResolveWriteIntentRange. Moreover, we additionally return numBytes
and resumeSpan in MVCCResolveWriteIntent and numBytes and resumeReason
in MVCCResolveWriteIntentRange, but these return values are currently
unused and serve as a placeholder in refactoring, but will be used in
the future.

Informs: #77228

Release note: None

rail committed:
logictest: unskip mixed version tests for ARM builds

The arm builds now use the expected tarball name format and can be used
in these tests.

This reverts commit df505daf1419968ed054d30f19f602101ebcb407.
Fixes: #94658

Epic: none
Release note: None

cucaroach committed:
sql: add telemetry for non-atomic copy usage

Keep account of customers switching back to non-atomic COPY, one day we
hope to remove non-atomic COPY so this is important data.

Release note: None

yuzefovich committed:
sql: switch from Compare to CompareError in some places

This commit switches some of the production usages of
`tree.Datum.Compare` method to `CompareError` in order to not crash the
server (when the vectorized engine is not used). With the vectorized
engine there was no user-visible impact we were catching the panic and
propagating it as a regular error already. Thus, I decided to not
include a release note.

Release note: None

yuzefovich committed:
tree: improve type-checking of ParenExprs

This commit makes it so that we strip parenthesis from `ParenExpr` when
type-checking a comparison op. We already do that (since 2017) in
`typeCheckComparisonOpWithSubOperator`, so it seems reasonable to also
do it in the other spot. This is needed in order for other logic (like
when handling tuple equality) to kick in.

Release note: None

yuzefovich committed:
roachtest: fix log message in multitenant/tpch

We had an unnecessary `%s` formatting directive there.

Epic: None

Release note: None

il9ue committed:
tree: allow multiple UDFs with in-exact type parameters

This patch makes it possible for udf to check
 exact type matching for udf parameters
Added TypeList.MatchExact to apply udf parameter check only.

Unit test for such family type parameter matching are provided

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95050)
<!-- Reviewable:end -->

il9ue committed:
tree: allow multiple UDFs with in-exact type parameters

This patch makes it possible for udf to check
 exact type matching for udf parameters
 Added TypeList.MatchExact to apply udf parameter check only.

 Unit test for such family type parameter matching are provided

 Release note: None

il9ue commented:
will create new PR
